src/librustc_mir/monomorphize/collector.rs

```diff
@@ -1264,7 +1264,13 @@ fn collect_const<'tcx>(
 ) {
     debug!("visiting const {:?}", constant);
 
-    match constant.val {
+    let substituted_constant = if let ConstValue::Param(param) = constant.val {
+        param_substs.const_at(param.index as usize)
+    } else {
+        constant
+    };
```

Comment:
IMO it would be easier to substitute in the caller or even here but irrespective of the kind of constant, right now Unevaluated also has unnecessarily specific code.

Not to mention this should probably also normalize.